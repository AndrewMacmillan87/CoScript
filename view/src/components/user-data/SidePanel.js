import React, { Fragment, useEffect, useState } from "react";
import { Link, withRouter } from "react-router-dom";

import { getUserMessages } from "../../redux/auxiliary/user-messages";

// Redux
import { connect } from "react-redux";

import PropTypes from "prop-types";

import "./side-panel.css";

export const SidePanel = withRouter(({ auth, ...props }) => {
    const [unreadMessages, setUnreadMessages] = useState(false);

    useEffect(() => {
        const check = async () => {
            const userMessages = await getUserMessages();
            for (let message of userMessages) {
                if (!message.read) {
                    setUnreadMessages(true);
                    return;
                }
            }
            setUnreadMessages(false);
        };
        check();
    }, []);

    const user = require("../../img/user.svg");

    const profile = (
        <aside className='side-panel-container'>
            <div className='profile-avatar-container'>
                <div>
                    <img
                        className='profile-avatar'
                        src={auth.user ? auth.user.avatar : user}
                        alt='Users profile'
                    />
                </div>
                <p className='avatar-caption'>
                    <span style={{ fontWeight: "bold" }}>
                        {auth.user ? auth.user.username : ""}
                    </span>
                </p>
            </div>
            <div className='profile-options-container'>
                <ul className='profile-options'>
                    <a href='#profile-h2'>
                        <li className='profile-choice'>Profile</li>
                    </a>
                    <a href='#photo-h2'>
                        <li className='profile-choice'>Photo</li>
                    </a>
                    <a href='#projects-h2'>
                        <li className='profile-choice'>My Projects</li>
                    </a>
                    {unreadMessages ? (
                        <div className='unread-message-link'>
                            <div className='unread-message-icon'></div>
                            <Link to='/messages'>
                                <li className='profile-choice'>Messages</li>
                            </Link>
                        </div>
                    ) : (
                        <Link to='/messages'>
                            <li className='profile-choice'>Messages</li>
                        </Link>
                    )}
                    <Link to='/account'>
                        <li className='profile-choice'>Account</li>
                    </Link>
                    <Link to='/'>
                        <li className='profile-choice'>Back Home</li>
                    </Link>
                    <a
                        onClick={() => {
                            props.history["from"] = "logout";
                            props.history.push("/redirect");
                        }}
                        href=' '>
                        <li className='profile-choice'>Sign Out</li>
                    </a>
                </ul>
            </div>
        </aside>
    );

    const account = (
        <aside className='side-panel-container'>
            <div className='profile-avatar-container'>
                <div>
                    <img
                        className='profile-avatar'
                        src={auth.user ? auth.user.avatar : user}
                        alt='Users profile'
                    />
                </div>
                <p className='avatar-caption'>
                    {" "}
                    <span style={{ fontWeight: "bold" }}>
                        {auth.user ? auth.user.username : ""}
                    </span>
                </p>
            </div>
            <div className='profile-options-container'>
                <ul className='profile-options'>
                    <Link to='/profile'>
                        <li className='profile-choice'>Profile</li>
                    </Link>
                    <a href='#account-h2'>
                        <li className='profile-choice'>Account</li>
                    </a>
                    <a href='#privacy-h2'>
                        <li className='profile-choice'>Privacy</li>
                    </a>
                    <Link to='/'>
                        <li className='profile-choice'>Back Home</li>
                    </Link>
                    <a
                        onClick={() => {
                            props.history["from"] = "logout";
                            props.history.push("/redirect");
                        }}
                        href=' '>
                        <li className='profile-choice'>Sign Out</li>
                    </a>
                </ul>
            </div>
        </aside>
    );

    const messages = (
        <aside className='side-panel-container'>
            <div className='profile-avatar-container'>
                <div>
                    <img
                        className='profile-avatar'
                        src={auth.user ? auth.user.avatar : user}
                        alt='Users profile'
                    />
                </div>
                <p className='avatar-caption'>
                    {" "}
                    <span style={{ fontWeight: "bold" }}>
                        {auth.user ? auth.user.username : ""}
                    </span>
                </p>
            </div>
            <div className='profile-options-container'>
                <ul className='profile-options'>
                    <Link to='/profile'>
                        <li className='profile-choice'>Profile</li>
                    </Link>
                    <Link to='/account'>
                        <li className='profile-choice'>Account</li>
                    </Link>
                    <Link to='/'>
                        <li className='profile-choice'>Back Home</li>
                    </Link>
                    <a
                        onClick={() => {
                            props.history["from"] = "logout";
                            props.history.push("/redirect");
                        }}
                        href=' '>
                        <li className='profile-choice'>Sign Out</li>
                    </a>
                </ul>
            </div>
        </aside>
    );

    switch (props.location.pathname) {
        case "/profile":
            return profile;
        case "/account":
            return account;
        case "/messages":
            return messages;
        default:
            return <Fragment />;
    }
});

SidePanel.propTypes = {
    auth: PropTypes.object,
};

const setProps = (state) => ({
    auth: state.auth,
});

export default connect(setProps, {})(SidePanel);
