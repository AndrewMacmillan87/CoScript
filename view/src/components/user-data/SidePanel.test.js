import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter as Router, Link } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import { Provider } from "react-redux";
import { reduxStore } from "../../redux/store/reduxStore";

import { SidePanel } from "./SidePanel";

describe("SidePanel component", () => {
    let props;
    beforeEach(() => {
        props = {
            auth: {
                user: {
                    avatar: "",
                    username: "test101",
                },
            },
            history: {
                from: "",

                push: jest.fn(),
            },
            location: {
                pathname: "",
            },
        };
    });

    afterEach(() => {
        props.location.pathname = "";
    });

    test("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(
            <Provider store={reduxStore}>
                <Router>
                    <SidePanel.WrappedComponent {...props} />
                </Router>
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it("renders correctly", () => {
        const tree = renderer
            .create(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should not render the component body when location.pathname is an empty string", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <SidePanel.WrappedComponent {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("aside").length).toEqual(0);
    });

    it("should render the component body when location.pathname is a valid path", () => {
        props.location.pathname = "/profile";
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <SidePanel.WrappedComponent {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("aside").length).toEqual(1);
    });

    it("should render a div element with a className of 'profile-avatar-container'", () => {
        props.location.pathname = "/profile";
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <SidePanel.WrappedComponent {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(".profile-avatar-container").length).toEqual(1);
    });

    it("should render an img element with a className of 'profile-avatar'", () => {
        props.location.pathname = "/profile";
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <SidePanel.WrappedComponent {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("img").length).toEqual(1);
        expect(wrapper.find(".profile-avatar").length).toEqual(1);
    });

    it("should render an p element with a className of 'avatar-caption'", () => {
        props.location.pathname = "/profile";
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <SidePanel.WrappedComponent {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("p").length).toEqual(1);
        expect(wrapper.find(".avatar-caption").length).toEqual(1);
    });

    it("should render an ul element with a className of 'profile-option'", () => {
        props.location.pathname = "/profile";
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <SidePanel.WrappedComponent {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("ul").length).toEqual(1);
        expect(wrapper.find(".profile-options").length).toEqual(1);
    });

    describe("profile view", () => {
        it("should render 7 li elements", () => {
            props.location.pathname = "/profile";
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("li").length).toEqual(7);
        });

        it("should render 3 Link components", () => {
            props.location.pathname = "/profile";
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find(Link).length).toEqual(3);
        });

        it("should render 7 a elements", () => {
            props.location.pathname = "/profile";
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("a").length).toEqual(7);
        });

        it("should call history.push on logout", () => {
            props.location.pathname = "/profile";
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            );
            wrapper.find("a").at(6).simulate("click");
            expect(props.history.push).toBeCalled();
        });
    });

    describe("account view", () => {
        it("should render 5 li elements", () => {
            props.location.pathname = "/account";
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("li").length).toEqual(5);
        });

        it("should render 2 Link components", () => {
            props.location.pathname = "/account";
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find(Link).length).toEqual(2);
        });

        it("should render 5 a elements", () => {
            props.location.pathname = "/account";
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("a").length).toEqual(5);
        });

        it("should call history.push on logout", () => {
            props.location.pathname = "/account";
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            );
            wrapper.find("a").at(4).simulate("click");
            expect(props.history.push).toBeCalled();
        });
    });

    describe("messages view", () => {
        it("should render 4 li elements", () => {
            props.location.pathname = "/messages";
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("li").length).toEqual(4);
        });

        it("should render 3 Link components", () => {
            props.location.pathname = "/messages";
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find(Link).length).toEqual(3);
        });

        it("should render 4 a elements", () => {
            props.location.pathname = "/messages";
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("a").length).toEqual(4);
        });

        it("should call history.push on logout", () => {
            props.location.pathname = "/messages";
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <SidePanel.WrappedComponent {...props} />
                    </Router>
                </Provider>
            );
            wrapper.find("a").at(3).simulate("click");
            expect(props.history.push).toBeCalled();
        });
    });
});
