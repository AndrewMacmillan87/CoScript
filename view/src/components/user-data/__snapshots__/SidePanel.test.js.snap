// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidePanel component renders correctly 1`] = `null`;
